lib/internal/http2.js

```diff
@@ -48,7 +46,139 @@ Object.defineProperty(exports, 'constants', {
 // emit events as callbacks. To allow this to happen,
 // it has to inherit from EventEmitter.
 util.inherits(http2.Http2Session, EventEmitter);
+util.inherits(http2.Http2Stream, Duplex);
 
+function maybeDestroyStream(stream) {
+  if ((!stream.readable &&
+       !stream.writable &&
+       !stream._writableState.length)) {
+    stream.destroy();
+  }
```

Comment:
Which `destroy`? On `stream` or on `_writableState`? The `stream.destroy()` method is exposed by `Http2Stream` in order to free up the underlying resources.

Some background: There are three key states that need to be tracked: (1) Is the Http2Stream attached to the underlying Http2Session (e.g. that is, is the internal nghttp2_stream handle still valid), (2) Is the Http2Stream Duplex interface still Writable, and (3) Is the Http2Stream Duplex still Readable. The Http2Stream object cannot be disposed of and garbage collected until all three of those states are false. Because of async reads/writes on the Duplex interface, it's possible for the Http2Stream to become detached from the underlying Http2Session. Ideally, when that happens both the Readable and Writable sides will close and any internally buffered data would be dropped and additional writes will fail but I do not have that completely wired up yet.

The `maybeDestroyStream(stream)` check is largely a temporary stopgap to determine when the Writable and Readable sides are closed so that we can allow the Http2Stream to be garbage collected.